core/parentchain/light-client/src/io.rs

```diff
@@ -103,78 +120,76 @@ where
 	// FIXME: That should be an unique path.
 	if SgxFile::open(LIGHT_CLIENT_DB).is_err() {
 		info!("[Enclave] ChainRelay DB not found, creating new! {}", LIGHT_CLIENT_DB);
-		return init_parachain_validator::<B, OCallApi>(params, ocall_api)
+		let validator = init_parachain_validator::<B, OCallApi>(
+			ocall_api,
+			RelayState::new(params.genesis_header, Default::default()).into(),
+		)?;
+		LightClientStateSeal::<B, LightValidationState<B>>::seal_to_static_file(
+			validator.get_state(),
+		)?;
+		return Ok(validator)
 	}
 
 	let (validation_state, genesis_hash) = get_validation_state::<B>()?;
 
-	let mut validator = init_parachain_validator::<B, OCallApi>(params.clone(), ocall_api)?;
-
-	if genesis_hash == params.genesis_header.hash() {
-		validator.set_state(validation_state);
-		// The init_parachain_validator function clear the state every time,
-		// so we should write the state again.
-		LightClientStateSeal::<B, LightValidationState<B>>::seal_to_static_file(
-			validator.get_state(),
-		)?;
+	let init_state = if genesis_hash == params.genesis_header.hash() {
 		info!("Found already initialized light client with Genesis Hash: {:?}", genesis_hash);
-	}
+		validation_state
+	} else {
+		info!(
+			"Previous light client db belongs to another parentchain genesis. Creating new: {:?}",
+			genesis_hash
+		);
+		RelayState::new(params.genesis_header, vec![]).into()
+	};
+
+	let validator = init_parachain_validator::<B, OCallApi>(ocall_api, init_state)?;
 	info!("light client state: {:?}", validator);
+
+	LightClientStateSeal::<B, LightValidationState<B>>::seal_to_static_file(validator.get_state())?;
 	Ok(validator)
 }
 
-fn get_validation_state<B: Block>() -> Result<(LightValidationState<B>, B::Hash)>
-where
-	B: Block,
-{
+// Todo: Implement this on the `LightClientStateSeal` itself.
```

Comment:
Just a confirmation: this `TODO` item is something to be tackled in an another PR, right?

Reply:
Yes, it is already in the working. :)